nexus/src/db/datastore/mod.rs

```diff
@@ -122,7 +122,7 @@ impl DataStore {
         self.pool.pool()
     }
 
-    pub(super) async fn pool_authorized(
+    pub async fn pool_authorized(
```

Comment:
I agree, it only looks like this is used for `no_disk_records_exist`.

Reply:
Okay, so this sorta sent me down a rabbit hole.

Between this, and the [comment](https://github.com/oxidecomputer/omicron/pull/1835#discussion_r1001936727) about the weird "sort-of-cyclic" relationship between `nexus` and `nexus-test-utils` below, I sorta tried to do a clean-up of test-helper code.

Basically: I wanted to be able to write test code that can do arbitrary DB queries, and call functions that let us poke into parts of Nexus' durable storage trivially. You bring up a great point, this interface being available for non-test purposes would be a clear break of functional boundaries, but IMO, requiring arbitrary, one-off Diesel queries to add boilerplate code to this interface would be a bummer.

So: to reconcile both these things, I'm making this a function that exists only within `#[cfg(tests)]`.

This *should* be doable. A function that exists for a unit-tests, but which does not exist beyond that functional boundary (so it can't be used incorrectly in a real prod usage).

*Should* be. Anyway, when I first tried just making that switch, it broke the tests I added to `disk_create.rs` - they were no longer able to access the test-only function, because they were actually integration tests - the version of [`omicron_nexus` being tested was actually the one exported from `nexus-test-utils`, which means it's not considered the same as the version of the *current* crate, also `omicron_nexus`, but built under #[cfg(test)]](https://github.com/oxidecomputer/omicron/pull/1835#discussion_r1001204291). When I originally tried writing the "saga tests" in `disk_create.rs`, I wanted them to be co-located with the source, to make it easier to eye-ball test coverage, but it was the first test in Nexus to actually use `#[nexus_test]`, which caused the cyclic dependency of `nexus-test-utils` -> `nexus` -> `nexus-test-utils`, which caused the error described [here](https://github.com/oxidecomputer/omicron/pull/1835#discussion_r1001936727). I wanted to re-use test setup code because my saga tests also wanted to validate state about simulated sled agents (and arguably in the future may want to check metrics information from Clickhouse).

To fix the tests, I had choices:
- **Make the `disk_create.rs` test dependencies much smaller, not including some of {CRDB, Clickhouse, Sled Agents}, such that test setup isn't needed**. This is possible, but it seems ill-advised -- the population step of Nexus is necessary for `authz` to function correctly, being able to poke at "sim sled agents" is necessary for all the Nexus HTTP requests to work (unless they could somehow be stubbed out, which seems like a pain), and doing so would also make it more difficult to monitor metrics later, if Clickhouse wasn't up.
- **Add interfaces to the datastore for test-only queries, keep this test as an integration test**. I didn't want to go down this route, because it seems like it would either bloat the interface with lower quality (intended for tests!) non-production code, or raise the bar on the safety/concurrency considerations of test-only queries (which would make tests more difficult to write).
- **Break the circular dependency the bad way**. Duplicate all the code Nexus used to do setup between unit and integration tests by copying everything we need from `nexus-test-utils` into `nexus`. This seems doable, but somewhat gnarly, especially since we're dealing with a lot of possible configurations and related services. Additionally, this feels a bit like "cause more problems for future us", which sucks.
- **Break the circular dependency in a better way**. Basically, make `nexus-test-utils` mostly no longer depend on `nexus`, such that the test actually *can* be a unit test, and so we actually *can* have `#[cfg(test)]` functions, like this one. I did this.

Anyway, this did end up requiring a fair bit more code changes, but it seems worth it. We can now write unit tests for Nexus that exist side-by-side with complex saga code, and we have the ability to write more powerful unit tests that can access the DB with ease. Furthermore, it's a step in the road for fixing a circular dependency in the build graph.

The change needed to make this all happen is described here: a0e52c99200542753387010bcd0af715deeca1d0

Reply:
I'm sorry for sending you down a rabbit hole!  Thanks for considering this problem.  I think you're right that it's important.

I'm a little on the fence about this solution.  This seems like a pretty big change _just_ to be able to get a database handle for the tests.  Do I understand right that there's only one impl of the new trait?  As a developer I find this pattern kind of annoying -- it feels like a bunch of extra cognitive overhead that's exposed to all the integration tests, plus all the test-utils code.  But maybe this is sufficiently hidden by the `type` definitions you added to each of the integration test files, and hopefully people won't really need to think about the trait very much.  I'm curious what anybody else thinks about this.  My avoidance of this pattern might be idiosyncratic.

Did you consider the approach of having a DatastoreTestInterfaces trait and putting `pool_for_tests` into that, similar to what we do for NexusTestInterfaces?  I think that trait exists for a very similar purpose and there's some benefit to using the same pattern, I think.  Although this is still a trait with only one impl, the only place that needs to know about it are the things that use this very specific interface (`pool_for_tests()`), as opposed to all the tests.

Reply:
To be clear, I'm _not_ saying this approach is a non-starter or anything, and if urgency is more important here, we can just go ahead with it!

Reply:
In multi-crate environments, I think it makes sense to think of `NexusServer` and `NexusServer`, but built with `#[cfg(test)]`, as two different structures. From the perspective of `omicron-nexus` using `nexus-test-utils` vs the integration tests using `nexus-test-utils`, those really are two different structures, even with a single "impl" body.

I actually do really like the `#[cfg(test)]` mechanism of wrapping test helpers, because of a few reasons:
- This code is never compiled into the *real* production binary, so it reduces the impact on the resulting binaries. If we want to add tons of diesel queries to test-only configurations, we 100% can, and the cost on the prod version of nexus should be non-existent.
- It makes a distinction between "interfaces usable by unit tests" vs "interfaces usable by integration tests". If integration tests need to be so invasive, they are probably better suited as integration tests.
- It makes it impossible to use these test interfaces from a non-test environment.

If we go down the route of "add a `DatastoreTestInterfaces`, revert e44438d85bd30d82fff99bd9b30b156289afef64", I think I'll run into a problem:

`nexus-test-utils` would still have a dependency on `omicron-nexus`, so there would still be the issue that unit tests for nexus cannot just `use crate::...` things, they would need to `use nexus_test_utils::re_exported_nexus::*` to be able to interoperate with the integration test setup.

I think this quirkiness, especially for unit tests, is the most confusing piece here. It threw me off quite a bit when writing these tests, and took a fair amount of digging to sort through.

If there *are* test interfaces that we want to expose to integration tests, I think we can. Nothing prevents us from both using a generic test framework, and *also* exposing a `DatastoreTestInterfaces` trait, but it seems preferable to limit that interface to "thing *truly* needed by integration tests, not just unit tests".